### ISSUE TITLE
terraform apply of large TypeSet is slow, assertSetValuesCompatible is too complex

### ISSUE THREAD

**Issue by freedge** [hashicorp/terraform#32940]
### Terraform Version

```shell
Terraform v1.5.0-dev
on linux_amd64
```

### Terraform Configuration Files

per https://github.com/PaloAltoNetworks/terraform-provider-panos/blob/b4ad451eb47b2c7e463d94b13fd7e2a5a62db158/docs/resources/address_objects.md 

```terraform
# Make address objects like "test1_1", "test1_2", ...
resource "panos_address_objects" "example" {
    dynamic "object" {
        for_each = setproduct(range(1, 6), range(1, 11))
        content {
            name = "test${object.value[0]}_${object.value[1]}"
            type = "ip-netmask"
            value = "10.${object.value[0]}.${object.value[1]}.0/24"
        }
    }

    lifecycle {
        create_before_destroy = true
    }
}
```

### Debug Output

```
...
2023-03-29T10:59:27.846+0200 [TRACE] checkPlannedChange: Verifying that actual change (action Update) matches planned change (action Update)
...
```

### Expected Behavior

terraform apply should be reasonably fast when applying object with large schema.TypeSet (8k objects)

### Actual Behavior

terraform apply is terribly slow

### Steps to Reproduce

unfortunately https://github.com/hashicorp/terraform-provider-null does not use schema.TypeSet so I don't know how to provide a reproducer

### Additional Context

https://github.com/hashicorp/terraform/blob/d9dfd451ea572219871bb9c5503a471418258e40/internal/plans/objchange/compatible.go#L347-L349

complexity is N^2

### References

- https://github.com/hashicorp/terraform/issues/32937#issuecomment-1488593598

**Comment by kmoe** [comment]
Thanks for the issue. Do you have an idea as to how that code could be improved?

**Comment by jbardin** [comment]
Yes, unfortunately this is inherently an n^2 operation because in order validate possible all changes in a set, each value might need to be compared with nearly every other value. I can see a little optimization to be had by not comparing values which have already been compared, which could be significant in some cases depending on how deep the values are structured.

The only real way though to not spend so much time validating these values is to not do it at all. That is a hard call however, because incorrect values here don't affect the resources causing the incorrect values, rather they cascade downstream to other resources which fail in confusing ways when their inputs change unexpectedly. Limiting the validation is something we have to look into more carefully.
